### DISCUSSION TITLE
Nvidia driver loading appears to be different in Cobia

### DISCUSSION THREAD

**Discussion by rowanmoul** [Jip-Hop/jailmaker#74] [Q&A]
In the main readme, [it suggests](https://github.com/Jip-Hop/jailmaker/blob/main/README.md#nvidia-gpu) creating a startup task to load the nvidia drivers, however the command given there is trying to run a binary that doesn't exist on my Cobia system (`/usr/bin/nvidia-modprobe`) . I don't have any bluefin systems to compare against, but I assume that binary existed on bluefin?
https://github.com/Jip-Hop/jailmaker/blob/56d53a1feb65e252aef06a4576b5f370ceabbb5f/README.md?plain=1#L143

That said, it appears that running `nvidia-modprobe` is not needed. If you load the kernel module (`modprobe nvidia-current-uvm`), the necessary device nodes will be created when jailmaker runs `nvidia-smi -f /dev/null` as part of the nvidia passthrough method on jail startup.
https://github.com/Jip-Hop/jailmaker/blob/56d53a1feb65e252aef06a4576b5f370ceabbb5f/jlmkr.py#L111

**Comment by Jip-Hop** [comment]
Looks like you got it working. Would you mind adding these instructions for Cobia and creating a pull request? I don't have a nvidia GPU so can't test this myself. Please also mention the version number besides Cobia in the readme. Thank you!

**Reply by rowanmoul** [reply]
I haven't actually got everything running, but I expect it to work since the device nodes are created in time for jailmaker to pass them through. The next step is to get docker running inside the jail and pass the gpu into the containers that will use it and verify that they are able to do so. I will do some more testing, and maybe I'll even install bluefin in a VM to compare against. It's entirely possible that `nvidia-modprobe` was never included in SCALE, but the startup task still did enough by loading the kernel module.

**Reply by rowanmoul** [reply]
I dug through the bluefin installer disk image and found the root filesystem squashfs. It does indeed include `/usr/bin/nvidia-modprobe`.
It looks like it is installable as a [separate package](https://packages.debian.org/bookworm/nvidia-modprobe), not bundled with the driver (this was the case under Debian Bullseye too), so I guess IX intentionally removed it for one reason or another.
I haven't had a chance to work on the rest of it just yet.

**Comment by Jip-Hop** [comment]
Could you please try [the jlmkr.py file from the develop branch](https://github.com/Jip-Hop/jailmaker/blob/develop/jlmkr.py)? I think it should now load the nvidia kernel module automatically.

**Reply by rowanmoul** [reply]
Sorry for the long delay, I haven't been able to put as much time into this as I would have liked but I finally got back to it. I can confirm that your changes in the develop branch are currently enough (in Cobia 2023.10.1.3) to load the drivers and pass the GPU through without any extra startup scripts.
One other thing that I found. If someone is passing through their Nvidia GPU so that they can use it in a docker container, they will need to install the `nvidia-container-toolkit` in the jail. I followed this documentation: https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/latest/install-guide.html#installing-with-apt
Following that, I installed version 1.14.4 of the toolkit (at the time of writing). The TrueNAS Scale version noted above has v1.13.4 installed, but since this isn't a driver component, I don't think it needs to perfectly match, [like was discovered here.](https://github.com/Jip-Hop/jailmaker/discussions/76#discussioncomment-8312178) If that ever changes, we may want to look at bind mounting the container toolkit components too.

While I am usually more comfortable in Arch Linux (that is where I first got started running Linux), I used a debian bookworm based jail for this so that (in theory) shared library versions would be in sync with those found on the host.
